### PR TITLE
Reintroduce changes for Bash with rework

### PR DESCRIPTION
**Pull request by lewis-yeung**
### Prerequisites

- [x] I have read and understood the [contributing guide][CONTRIBUTING.md].
- [x] The commit message follows the [conventional commits][cc] guidelines.
- [ ] Tests for the changes have been added (for bug fixes / features).
- [x] Docs have been added/updated (for bug fixes / features).

### Description

~Sadly, I failed to find a `preexec`-like solution for Bash other than the original `PS0` trick. It's worth noting that overriding the `DEBUG` trap in Bash is more hacky and will break something like [bash-preexec][bash-preexec]. So we should always enable the `promptvars` shell option as opposed to the previous buggy implementation, and escape special characters in a prompt string.~

Finally, with some rework and careful testing, we can safely reintroduce two previously reverted changes (in #5473).

[CONTRIBUTING.md]: https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/CONTRIBUTING.md
[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary
[bash-preexec]: https://github.com/rcaloras/bash-preexec

### CODE REVIEW

**Comment by lewis-yeung** [comment]
This time, to avoid introducing other bugs, I'll further review the whole prompt logic before the changes are merged.

File: src/terminal/writer.go [hunk 1]
```diff
@@ -392,17 +400,18 @@ func write(s rune) {
...
-	if !Interactive {
-		for special, escape := range formats.EscapeSequences {
-			if s == special && lastRune != escape {
-				builder.WriteRune(escape)
-			}
+	// UNSOLVABLE: When "Interactive" is true, the prompt length calculation in Bash/Zsh can be wrong, since the final string expansion is done by shells.
+	length += runewidth.RuneWidth(s)
+	// length += utf8.RuneCountInString(string(s))
+
+	if !Interactive && !Plain {
```
Comment (lewis-yeung) [inline src/terminal/writer.go:407]: A problem here: when `Interactive` is `true`, the escape will not be done by OMP here, however, a series of backslashes will be half eaten by the shell escape logic.

File: src/shell/formats.go [hunk 2]
```diff
@@ -50,9 +50,10 @@ func GetFormats(shell string) *Formats {
 			ITermPromptMark:       "\\[$(iterm2_prompt_mark)\\]",
 			ITermCurrentDir:       "\\[\x1b]1337;CurrentDir=%s\x07\\]",
 			ITermRemoteHost:       "\\[\x1b]1337;RemoteHost=%s@%s\x07\\]",
-			EscapeSequences: map[rune]rune{
-				96: 92, // backtick
-				92: 92, // backslash
+			EscapeSequences: map[rune]string{
+				'\\': `\\\\`,
+				'`':  "\\`",
+				'$':  `\$`,
```
Comment (lewis-yeung) [inline src/shell/formats.go]: These two characters should be escaped regardless of `terminal.Interactive`. 🤦🏻

**Comment by lewis-yeung** [comment]
Things always go tricky when falling on Bash...

**Comment by lewis-yeung** [comment]
GOOD NEWS! Found a solution where we need only a simple trick on the shell side, without handling the annoying escape in OMP. Will update the PR later.

**Comment by lewis-yeung** [comment]
Ready for review now.

**Comment by JanDeDobbeleer** [comment]
@lewis-yeung looking at this one when I get back from holidays.

**Pull request closed without merging**